qemu/target-mips/translate.c

```diff
@@ -2551,7 +2551,7 @@ static void gen_logic_imm(DisasContext *ctx, uint32_t opc,
             tcg_gen_ext32s_tl(tcg_ctx, *cpu_gpr[rt], *cpu_gpr[rt]);
             MIPS_DEBUG("aui %s, %s, %04x", regnames[rt], regnames[rs], imm);
         } else {
-            tcg_gen_movi_tl(tcg_ctx, *cpu_gpr[rt], uimm << 16);
+            tcg_gen_movi_tl(tcg_ctx, *cpu_gpr[rt], imm << 16);
```

Comment:
As previous `(rs != 0 && (ctx->insn_flags & ISA_MIPS32R6))` infers, only `MIPS32R6` instruction set requires sign extension and the logic here implements it correcly.

Reply:
@wtdcode , this instruction should always perform sign extension - you can check it in the modern QEMU: it uses 'imm', not 'uimm', in these lines; the problem of 'uimm' is that 'uimm << 16' is an unsigned 32-bit value, and it is cast (unsigned-extended) to 64-bit, but this is incorrect and causes errors during emulation.

Reply:
Sorry I did a double-check and you are right indeed.